src/System.Private.ServiceModel/src/Internals/System/Runtime/AsyncLock.cs

```diff
@@ -135,7 +165,23 @@ public void Dispose()
                 _asyncLock._lockTakenCallStack = null;
                 _asyncLock._lockTakenCallStackString = null;
 #endif
-                _asyncLock._lockTaken.Value = false;
+                if (s_heldLocks.Value == _asyncLock) // This is the only lock entered
+                {
+                    s_heldLocks.Value = null;
+                }
+                else if (s_heldLocks.Value is List<AsyncLock> listOfLocks)
+                {
+                    Fx.Assert(listOfLocks.Contains(_asyncLock), "The list of AsyncLock's didn't contain the expected lock");
+                    // As locks are expected to be released in the order they are taken and they are always appended to the end,
+                    // removal should be O(n) simply to look for the lock and removal should be constant time. If this becomes
+                    // a significant overhead, then manual search in reverse will fix it. Keeping simple for now.
+                    listOfLocks.Remove(_asyncLock);
```

Comment:
What makes this thread-safe?  If I start multiple async operations from the same context, won't they all end up potentially operating on the same list, and potentially trying to remove from the same list in parallel?